### ISSUE TITLE
Animation frame rate

### ISSUE THREAD

**Issue by gentlegiantJGC** [GSD6338/XmasTree#2]
Looking through the running code it is depending on the delay of the show method to do frame delay.

Is there a known frame rate that the tree will run at? It looks like this will vary based on hardware speed.

I can look into submitting code to at least making it a known frame rate (less than the hardware max)

I could also look into making the frame rate configurable but I don't know how you would want this done. Perhaps as an extra optional column that defaults to 1/60th of a second. Is the tree capable of running at this speed?

**Comment by melroy89** [comment]
[The code says](https://github.com/GSD6338/XmasTree/blob/main/03_execution/run.py#L15):

```py
sleep_time = 0.017  # approx 60fps
```

So I guess 60 FPS?


That being said, this `sleep_time` is not used. It's commented out!?

**Comment by gentlegiantJGC** [comment]
Yes this is why I was asking.

I think there is a fairly large delay in the show method that slows it down quite a lot.

The better way of doing it is getting the time at the start of the frame, processing the frame and at the end calculating the amount of time needed to sleep to make it up to the end of that frame.

**Comment by gentlegiantJGC** [comment]
I have modified the run.py file to make these changes and a bunch of cleanup and other improvements.

I need to have another look over it but the code is here. I will open a pull request when it is ready.
https://github.com/gentlegiantJGC/XmasTree/blob/main/03_execution/run.py

**Comment by PhilippDahlinger** [comment]
I approve the changes from gentlegiantJGC,  I especially like the idea of varying delay between frames, this will help making animations simpler. Really hope this pull request will be accepted!

**Comment by BoostCookie** [comment]
The problem still is that the framerate info is not part of the csv-file. But better to have a hardcoded standard than no standard at all.

**Comment by gentlegiantJGC** [comment]
My suggested changes adds support for an optional FRAME_TIME column which specifies how long that frame should be displayed for.

If that column is not defined it will default to 1/60th of a second

The code that extracts frame time from the CSV is here
https://github.com/gentlegiantJGC/XmasTree/blob/main/03_execution/run.py#L69

**Comment by garciadelcastillo** [comment]
Hi there!

We looked into the issue of framerate but, as it turns out, any form of introduced pause via `sleep_time` would also introduce a very noticeable "lag" on the tree that would make make the animation not fluid at all. So, even though theoretically it would be great to control the fps, we hadn't found a good way of doing it hardware-wise. 

Our experience is that my RPi 4 could run animations at roughly 60 fps. 

I support adding a `FRAME_RATE` column in the CSV. Just put it all the way as the last column, to maintain compatibility with older code that reads columns by order, not by header name. 

You may want to submit a PR with changes added to `run-folder.py` as well

**Comment by gentlegiantJGC** [comment]
Yeh I was planning to modify that one as well.

Could you elaborate on what you mean by lag?
In the original run code it looked like you were experimenting with delay but you were just hard coding the sleep time in addition to the unknown delay in calling the neopixel show method.
I am getting the time at the start and end of frame processing how long to sleep for to make it up to the full frame time.

You should probably add an official document outlining the format. The best I could see was in the description of Matt's video and based on the examples. Only upon reading the code did I realise the order limitations.

Edit: just seen you called it FRAME_RATE not FRAME_TIME. I was expecting that column to store the time in seconds that that frame would be displayed before moving to the next frame

**Comment by Diggsey** [comment]
@garciadelcastillo is there video of Matt running an example CSV on his tree using the "player" program? That way we can figure out exactly how long each frame is.

**Comment by garciadelcastillo** [comment]
@gentlegiantJGC yup, documentation is nice, you are welcome to submit a draft with your PR if you fancy :)

@Diggsey there is a video of sample code ran on the tree in the ParametricCamp mini series linked on the README.

**Comment by garciadelcastillo** [comment]
@gentlegiantJGC feel free to add a column in position [1] (we start counting with zero) named `FRAME_TIME` that indicates the length of that frame in milliseconds. This would also be nice if people start making "variable rate" animations. Also, please make sure the code is updated to reflect this new column or the absence thereof (for legacy files).

**Comment by gentlegiantJGC** [comment]
> feel free to add a column in position [1]

Do you mean to the right of the FRAME_ID column or replacing that column?

I think if it was added in addition to the FRAME_ID column it would break the old code since the old code assumes that the colours start in the second column.

Replacing the FRAME_ID would work because it isn't used by the old code but that would remove an old feature which I don't like.

I think adding it to the end would be the best bet.

**Comment by garciadelcastillo** [comment]
Whatever you decide for the PR, it should leave the code working.

**Comment by Diggsey** [comment]
@garciadelcastillo Based on the video from https://youtu.be/ma4JF6b4Dqk?t=246 to https://youtu.be/ma4JF6b4Dqk?t=322 you run through 2638 frames in 76 seconds, meaning you get about **34.7 frames per second.**

No idea if Matt will get the same framerate, but I guess it shouldn't be too far off.

**Comment by garciadelcastillo** [comment]
Hahaha, that's very awesome @Diggsey , thanks for the timer lol. I guess it is about 30fps then! 

I believe Matt uses a RPi Zero, so maybe he has more latency? No idea.

**Comment by gentlegiantJGC** [comment]
So I can do one of two things if the frame time is not defined
1) hard code at a fixed FPS 30/60?
2) run at max hardware speed

**Comment by gentlegiantJGC** [comment]
I think we should default to 30fps. That will enable everyone to standardise their animations.
I will still add the FRAME_TIME column to allow it to be configured.

**Comment by gentlegiantJGC** [comment]
Just had another look over the old code and realised that this line stops us being able to add any columns to the format
https://github.com/GSD6338/XmasTree/blob/main/03_execution/run.py#L11

The new code is able to read the old and new formats just fine because it reads the column headers but the old code has a number of limitations.

The options to add a FRAME_TIME column are
1) Replace the FRAME_ID colum
2) Break the old python running code

Option 2 is the best option in my opinion since the old code has a number of issues and supporting it will severly limit the format. Users will just have to update to the new code.

**Comment by gentlegiantJGC** [comment]
Just spotted that you have some other files (grashopper) that also parse this format.
Do they use the header names or assume everything is positional?

**Comment by garciadelcastillo** [comment]
Hi @gentlegiantJGC, I propose:

- Animation is run default at max hardware speed (no `sleep`) if no frame rate information is present. My tests using `sleep` yielded really laggy results on the tree.
- You are welcome to add FRAME_TIME as the second column, but all the python code should be updated to reflect this change, as well as incorporate sanity checks in case the column is missing (legacy files). 
- The Grasshopper files use positional columns, but I am happy to fix those once the new column is added.
- No PRs that break old code will be accepted (except for the above).

**Comment by garciadelcastillo** [comment]
I would also be ok replacing the FRAME_ID column with the time one: I don't think the ID is used in any code, and it is kind of redundant.

**Comment by gentlegiantJGC** [comment]
Okay that sounds good to me. I will update the code and format doc.

My code does not have any positional requirements and will populate any missing column with default data so technically any column could be missing or in any order and the code will read it just fine.

The issue is any old code that assumes it is positional

**Comment by gentlegiantJGC** [comment]
I have just come across the blending code in run-folder.py
I don't know how this would work with a variable framerate.
Is it alright if it just jumps from the end of one sequence to the start of the next?
The alternative is to blend from the last frame to the first frame of the next.

It would be a shame to remove that code but I don't know how to make it work.

**Comment by Diggsey** [comment]
@gentlegiantJGC You could also interpolate the frame times, so if it goes from a slow playing sequence to a fast playing sequence it would gradually speed up, and vice versa.

Given that the purpose is just to blend between animations, it doesn't need to perfectly reproduce the animation of either for that transition period.

**Comment by gentlegiantJGC** [comment]
The issue is that each frame can specify how long it takes.

In the end I am checking if the last and first frames are substantionally different and if they are interpolating between just those frames. It isn't the same as the original but it works and also doesn't interpolate if the frames were similar enough (useful for animation files that are designed to loop otherwise there would be a weird pause as it restarts)

**Comment by gentlegiantJGC** [comment]
My changes are ready to review in #4
